R/nefsc_groundfish_access.R

```diff
@@ -955,9 +955,126 @@ add_area_stratification <- function(survdat_weights, include_epu = F, box_locati
+#'
+#' @description make a tidy occupancy dataset from the cleaned SURVDAT dataset, where each row represents a unique tow - species - occupancy record for every species in species_keep. In creating this occupancy dataset, we take the total biomass or abundance across sex/sizes and impute "absence" observations, such that there is a record for every species in `species_keep` at every unique tow.
+#'
+#' @param survdat_clean Survdat data, after usual preparations are completed.
+#' These include removal of old strata, labeling of areas of interest, and inclusion
+#' of the annual effort in each.
+#' @param species_keep Vector of character species names or vector of numeric species codes to filter SURVDAT['comname'] (if character), or SURVDAT['svspp'] (if numeric)
+#'
+#' @return tidy occupancy dataframe containing one row for every unique tow-species
+#' @export
+#'
+#' @examples
+#' # not run
+#' # make_survdat_occu(survdat_clean = survdat_clean, species_keep = c("atlantic cod", "offshore hake"))
+make_survdat_occu <- function(survdat_clean, species_keep){
+
+
+  #### 1. Filter SURVDAT to species of interest based on character or numeric vector `species_keep` and aggregate abundance and biomass data for these species at each tow location   ####
+  if (all(is.character(species_keep))) {
+    presence_data <- survdat_clean %>%
+      dplyr::filter(., comname %in% species_keep) %>%
+      dplyr::group_by(., id, svspp, comname) %>%
+      dplyr::summarise(
+        sum_abundance = sum(abundance),
+        sum_biomass_kg = ifelse(length(unique(biomass_kg == 1)), unique(biomass_kg), sum(biomass_kg))
+      ) %>%
+      dplyr::mutate(presence = ifelse(sum_abundance > 0, 1, 0)) %>% # should all be 1s
+      # presence = 1 if abundance >=1, presence = 0 if abundance = 0
+      dplyr::select(id, svspp, comname, presence, sum_abundance, sum_biomass_kg) %>%
+      dplyr::ungroup()
+  } else {
+    presence_data <- survdat_clean %>%
+      dplyr::filter(., svspp %in% species_keep) %>%
+      dplyr::group_by(., id, svspp, comname) %>%
+      dplyr::summarise(
+        sum_abundance = sum(abundance),
+        sum_biomass_kg = ifelse(length(unique(biomass_kg == 1)), unique(biomass_kg), sum(biomass_kg))
+      ) %>%
+      dplyr::mutate(presence = ifelse(sum_abundance > 0, 1, 0)) %>% # should all be 1s
+      # presence = 1 if abundance >=1, presence = 0 if abundance = 0
+      dplyr::select(id, svspp, comname, presence, sum_abundance, sum_biomass_kg) %>%
+      dplyr::ungroup()
+  }
+  
+  #### 2. Create dataframe with all possible tow/species combinations   ####
+  # Create a dataframe of all possible survey ID/species combinations
+  all_ID_SPEC_possibilities <- tibble::tibble(id = rep(unique(survdat_clean$id), length(unique(presence_data$svspp)))) %>%
+    dplyr::mutate(svspp = rep(unique(presence_data$svspp), length(unique(survdat_clean$id))),
+    comname = rep(unique(presence_data$comname), length(unique(survdat_clean$id)))) %>%
+    dplyr::arrange(id)
```

Comment:
Think you could accomplish this with:
```
all_spp <- survdat %>% distinct(comname, svspp)
all_id_spec_possibilites <- survdat_clean %>% tidyr::expand(id, comname) %>% left_join(all_spp, by = "comname")
```

Reply:
can't expand on both svspp and comname or it will mix and match the unique combinations, that is why I used a left_join as a second step to bring the svspp codes over while preserving their matchups

Reply:
```
As this continues on we do:
survdat_occu<- all_ID_SPEC_possibilities %>% 
    dplyr::left_join(presence_data, by = c("id", "svspp", "comname"))
```
I think that would add in an observation for every single species (400+) recorded by the trawl survey at each tow location? Could I fix that by changing `all_spp` to only be the ones we are interested in?

Reply:
I think got an option that will work...

```
 all_id_spec_possibilites <- survdat_clean %>%
   tidyr::expand(id, comname) %>%
   left_join(all_spp, by = "comname") %>%
   dplyr::filter(., comname %in% presence_data$comname)
 survdat_occu<- all_id_spec_possibilites %>% 
    dplyr::left_join(presence_data, by = c("id", "svspp", "comname"))
```